internal/overlord/cmdstate/cmdmgr.go

```diff
@@ -0,0 +1,619 @@
+		}
+		ptys = append(ptys, stdoutReader)
+		ttys = append(ttys, stdoutWriter)
+		stdout = stdoutWriter
+		wgEOF.Add(1)
+		go func() {
+			<-wsutil.WebsocketSendStream(ioConn, stdoutReader, -1)
+			stdoutReader.Close()
+			wgEOF.Done()
+		}()
+
+		// Receive from cmd.Stderr pipe, write to "io" websocket as well (or
+		// "stderr" websocket if client wants stderr separate).
+		stderrReader, stderrWriter, err := os.Pipe()
+		if err != nil {
+			return err
+		}
+		ptys = append(ptys, stderrReader)
+		ttys = append(ttys, stderrWriter)
+		stderr = stderrWriter
+		stderrConn := ioConn // TODO(benhoyt): this won't work as is -- websocket.Conn writes aren't concurrency safe
+		if s.stderr {
+			s.connsLock.Lock()
+			stderrConn = s.conns[wsStderr]
+			s.connsLock.Unlock()
+		}
+		wgEOF.Add(1)
+		go func() {
+			<-wsutil.WebsocketSendStream(stderrConn, stderrReader, -1)
+			stderrReader.Close()
+			wgEOF.Done()
+		}()
+	}
+
+	finisher := func(cmdResult int, cmdErr error) error {
+		for _, tty := range ttys {
+			tty.Close()
+		}
+
+		s.connsLock.Lock()
+		conn := s.conns[wsControl]
+		s.connsLock.Unlock()
+
+		if conn == nil {
+			if s.terminal {
+				controlExit <- true
+			}
+		} else {
+			conn.Close()
+		}
+
+		close(attachedChildIsDead)
+
+		wgEOF.Wait()
+
+		for _, pty := range ptys {
+			pty.Close()
+		}
+
+		setApiData(change, cmdResult)
+
+		return cmdErr
+	}
+
+	if s.timeout != 0 {
+		var cancel func()
+		ctx, cancel = context.WithTimeout(ctx, s.timeout)
+		defer cancel()
+	}
+
+	cmd := exec.CommandContext(ctx, s.command[0], s.command[1:]...)
+
+	// Prepare the environment
+	for k, v := range s.env {
+		cmd.Env = append(cmd.Env, fmt.Sprintf("%s=%s", k, v))
+	}
+
+	cmd.Stdin = stdin
+	cmd.Stdout = stdout
+	cmd.Stderr = stderr
+
+	cmd.SysProcAttr = &syscall.SysProcAttr{}
+	if s.uid != nil && s.gid != nil {
+		cmd.SysProcAttr.Credential = &syscall.Credential{
+			Uid: uint32(*s.uid),
+			Gid: uint32(*s.gid),
+		}
+	}
+
+	// Creates a new session if the calling process is not a process group leader.
+	// The calling process is the leader of the new session, the process group leader of
+	// the new process group, and has no controlling terminal.
+	// This is important to allow remote shells to handle ctrl+c.
+	cmd.SysProcAttr.Setsid = true
+
+	// Make the given terminal the controlling terminal of the calling process.
+	// The calling process must be a session leader and not have a controlling terminal already.
+	// This is important as allows ctrl+c to work as expected for non-shell programs.
+	if s.terminal {
+		cmd.SysProcAttr.Setctty = true
+	}
+
+	cmd.Dir = s.cwd
+
+	err = cmd.Start()
+	if err != nil {
+		return finisher(-1, err)
+	}
+
+	if s.terminal {
+		attachedChildIsBorn <- cmd.Process.Pid
+	}
+
+	err = cmd.Wait()
+	if err == nil {
+		return finisher(0, nil)
+	}
+
+	if errors.Is(ctx.Err(), context.DeadlineExceeded) {
+		return finisher(-1, fmt.Errorf("timed out after %v", s.timeout))
+	}
+
+	exitErr, ok := err.(*exec.ExitError)
+	if ok {
```

Comment:
inline assignment/declaration here, then make them all else if blocks, reads easier to see the error is being handled.

Then move the success case as the final return.

Reply:
Done.